feat: index historical votes

wgwz opened the pull request:
Closes: #22 
Closes: #37 (I decided to add the fix for this here, since it's small and potentially needed in this PR)

wgwz commented on sql/V1_12__votes.sql:
```diff
@@ -0,0 +1,39 @@
+DROP CONSTRAINT IF EXISTS proposals_proposal_id_ux;
+
+ALTER TABLE IF EXISTS proposals
+ADD CONSTRAINT proposals_proposal_id_ux UNIQUE (chain_num, proposal_id);
```
i needed to add this unique constraint on the proposals table because i wanted to utilize that column combination as a foreign key in the votes table. this provides at least two benefits, data integrity, in the sense that we won't add votes to the database for non-existent proposals, and also it provides a relationship that can be conveniently traversed in a postgraphile graphql query.

i.e. return all proposals and include the votes for each proposal:

```graphql
{
  allProposals(orderBy: PROPOSAL_ID_DESC, condition: {chainNum: 84}) {
    nodes {
      txHash
      proposalId
      votesByChainNumAndProposalId {
        nodes {
          voter
          option
        }
      }
    }
  }
}
```

wgwz commented on index_votes.py:
```diff
@@ -0,0 +1,118 @@
+                        logger.debug(_cur.statusmessage)
+                        pg_conn.commit()
+                        logger.info("vote inserted..")
+                    except ForeignKeyViolation as exc:
```
this exception block handles the scenario where a proposal has not yet been pruned, but a vote has been cast. once that proposal is pruned and the indexer has recorded it in the proposals table, then the next time the indexer checks for votes, it will pick these up and insert them into the database for the proposal.

in other words, until we've indexed the proposal, we will not index the votes. the proposal is not considered historical until it is pruned, and so it does seem to make sense. votes are only considered historical once the proposal becomes historic.

that said, we can change this behavior, however if we do, we lose the ability to traverse the relationship via graphql as mentioned above (since we would need to drop the foreign key).

wgwz replied:
> So currently we are fetching all votes each time a single vote is submitted, looping through the votes each time we fetch all votes, and then rolling back each commit if the commit fails due to a foreign key violation (i.e. the proposal has not yet been indexed because we have a constraint in place that requires the proposal to exist).

this works as follows:
1. find all `EventVote`'s that we have not yet indexed.
2. loop through those votes
3. for each of those votes, attempt to insert and commit the vote into the database 
4. if we encounter a foreign key error, it means the corresponding proposal has not yet been indexed
5. we only rollback the transaction for a single vote if 4 is hit

So votes that have their corresponding proposals in the database will indeed be committed.

> Wouldn't this require an EventVote to be emitted after the proposal is indexed but a proposal is only indexed after the proposal is pruned and the voting period has ended? 

It works because the process will always attempt to index votes which have not been successfully indexed.
We will keep trying to index votes over and over again, until it succeeds.
Until the proposal is pruned, we might have a queue of votes waiting for indexing.
As soon as the proposal is pruned and exists in the indexer database, all of the votes for that proposal will be indexed in the next run of the index votes process.

ryanchristo replied:
Are we attempting to iterate through all votes from the response even after the first vote was not stored due to the foreign key violation with the constraint? I tested with a single member and vote but haven't tested with multiple yet.

ryanchristo replied:
This approach is a bit concerning looking at the database logs:
https://github.com/regen-network/indexer/actions/runs/6042565087/job/16397918758?pr=41#step:5:88

We are attempting to index votes every second until the proposal has been indexed...

ryanchristo replied:
Also, is this process stuck on an event for all groups and all proposals? For example, I have a group proposal in one group with a 2 week voting period and a group proposal in another group with a 1 day voting period. If I submit a vote to the group proposal with a 2 week voting period, then a vote to the group with a 1 day voting period, will my vote not be indexed for the group with the 1 day voting period until the first proposal has been indexed?

ryanchristo replied:
I can try to test this out and answer my own questions but let me know if you have any thoughts in the meantime.

wgwz replied:
> This approach is a bit concerning looking at the database logs:
> https://github.com/regen-network/indexer/actions/runs/6042565087/job/16397918758?pr=41#step:5:88
> 
> We are attempting to index votes every second until the proposal has been indexed...

So we can adjust the timing...? I'd be concerned if we had a proven issue with cpu/memory in the database

wgwz replied:
>Also, is this process stuck on an event for all groups and all proposals? For example, I have a group proposal in one group with a 2 week voting period and a group proposal in another group with a 1 day voting period. If I submit a vote to the group proposal with a 2 week voting period, then a vote to the group with a 1 day voting period, will my vote not be indexed for the group with the 1 day voting period until the 2 week proposal has been indexed?

No, votes are only blocked if their corresponding proposal has not been pruned.

wgwz replied:
> Are we attempting to iterate through all votes from the response even after the first vote was not stored due to the foreign key violation with the constraint? I tested with a single member and vote but haven't tested with multiple yet. We should make sure we are exiting the loop as all other votes for the proposal would fail and rollback.

Makes sense, we can do that easily.

wgwz replied:
> I tested with a single member and vote but haven't tested with multiple yet. 

I think I may have a bug in the way the indexer works for multiple voters.
I'll look into next week.
But basically I think I missed the fact that this API will return all the votes on the proposal including the voter for a particular `EventVote`:
/cosmos/group/v1/votes_by_proposal/{proposal_id}
So I think we'd end up with duplicates in the database with the current code.

I'll figure this out next week and do some more testing.
Thanks for mentioning multiple voters @ryanchristo .

ryanchristo replied:
Yea, just posted this at the same time: https://github.com/regen-network/indexer/pull/38#discussion_r1312389707

wgwz commented on index_votes.py:
```diff
@@ -0,0 +1,118 @@
+def fetch_votes_by_proposal(height, proposal_id):
+    resp = requests.get(
+        f"{os.environ['REGEN_API']}/cosmos/group/v1/votes_by_proposal/{proposal_id}",
+        headers={"x-cosmos-block-height": str(height)},
```
note: we will hit this issue in redwood:
https://github.com/regen-network/indexer/issues/37

blushi commented:
> while testing this historical votes piece a bit further in my local environment, i'm observing that proposals that are rejected, do not get saved by the indexer in our proposals table. this seems to be because `cosmos.group.v1.EventProposalPruned` does not get emitted for rejected proposals. is that intended? am i wrong?
> 
> for that reason, historical rejected proposals are currently not supported with the indexer. consequently, we are unable to store historical votes for rejected proposals. is this a problem we need to solve in a separate issue? /cc @blushi @ryanchristo

Could you tell how you tested this in details? I believe such proposals should get pruned eventually and emitting `cosmos.group.v1.EventProposalPruned` at voting period + max exec period: https://github1s.com/cosmos/cosmos-sdk/blob/HEAD/x/group/keeper/keeper.go#L370

wgwz commented:
> Could you tell how you tested this in details? I believe such proposals should get pruned eventually and emitting `cosmos.group.v1.EventProposalPruned` at voting period + max exec period: https://github1s.com/cosmos/cosmos-sdk/blob/HEAD/x/group/keeper/keeper.go#L370

thanks @blushi i think the max exec period explains why i'm seeing this. it looks like that defaults to two weeks. i only created the rejected proposal a couple days ago, so that would explain it for my case.

wgwz commented:
> i'm going to re-test this with an example @ryanchristo gave me where you can lower the exec time

so this example was just about adjusting the min exec time.
as mentioned in stand-up today, we'd have to customize the max exec time locally (by [adjusting this value][1] if i'm not mistaken), so i'll just assume it works and move on.

that said, it's something to be aware of when testing the contents of the indexer database.
a rejected proposal and it's votes will not appear in the database until max exec period has passed.

[1]: https://github1s.com/cosmos/cosmos-sdk/blob/HEAD/x/group/config.go#L17-L18

ryanchristo reviewed (commented):
Looking good. Haven't tested yet. A couple questions/comments.

ryanchristo commented on utils.py:
```diff
@@ -145,3 +146,12 @@ def events_to_process(cur, index_table_name):
         yield list(g)
+
+
+def is_archive_node():
+    # since the indexer is intended to run against archive nodes,
+    # assume that by default the node is an archive node.
+    value = os.environ.get("REGEN_IS_ARCHIVE_NODE", "true").lower()
+    if value not in ["true", "false"]:
+        raise ValueError("REGEN_IS_ARCHIVE_NODE must be true or false")
+    return value == "true"
```
We also might want to assume archive node (i.e. the default if environment variable not provided) given this is how the indexer is intended to be used and a non-archive node is the exception (i.e. setting variable is required).

ryanchristo replied:
I'm ok with no changes but thinking in terms of minimal setup and room for error in production.

wgwz replied:
my thinking here was that by default, the indexer is meant to be run against an archive node. so in production, we wouldn't actually have to set this variable at all. and this variable only needs to be set when running against non-archive nodes.

ryanchristo commented:
Successfully tested the vote is not stored if the proposal has not yet been indexed and the vote is stored after the proposal is indexed: https://github.com/regen-network/indexer/actions/runs/6042565087/job/16397918758?pr=41

ryanchristo commented:
In addition to adding comments, can you make sure we add `V1_12__votes.sql` to the all migrations script?

ryanchristo commented on sql/V1_12__votes.sql:
```diff
@@ -0,0 +1,39 @@
+    OPTION TEXT NOT NULL,
+    metadata TEXT NOT NULL,
+    submit_time timestamptz NOT NULL,
+    PRIMARY KEY (
+      chain_num,
+      block_height,
+      tx_idx,
+      msg_idx
+    ),
```
So this primary key is not going to work when there are two votes. I can share a link here in a second but two votes for the same proposal hits a duplicate key violation error.

ryanchristo replied:
I think we just need to add the voter.

wgwz replied:
Thanks, I'll take a better look next week, oversights on my part with multiple voters.

ryanchristo replied:
Here's the link: https://github.com/regen-network/indexer/actions/runs/6043549097/job/16400735995#step:5:639

ryanchristo replied:
The db logs get pretty nasty when this occurs so may have trouble loading the db logs.

![Screenshot from 2023-08-31 17-01-45](https://github.com/regen-network/indexer/assets/12519942/cbf13509-180d-4258-a013-1a5148eb48fd)

wgwz commented:
with [the latest commit](https://github.com/regen-network/indexer/pull/38/commits/6f29a66fea01a7999d3a5a3e03b17edb445a5c56) i've changed the architecture of the index votes process to solve a few problems.
most importantly, properly handling the fact that we need to get all votes for a proposal with the current implementation of `EventVote` (which only specifies `proposal_id` and not the voter and option).

i've also dropped the relationship between the proposals table and the votes.
this just means we can't traverse that relationship in graphql as i had mentioned in a previous comment.
that should be fine, we'll still have the ability to write graphql that queries the votes table.

last but not least i added a new function for figuring out which events to process (`new_events_to_process`).
this is a slightly different version of the already existing `events_to_process` function.
the difference being that instead of finding **all** `msg_events` of a particular type that have yet to be indexed, it finds only the newest events that occurred since the last time it ran.

`events_to_process` returns absolutely all events that have not been processed.
this is great when you want to have all historical data at every point in time (i.e. retirements (might not be the best example)).
but it's not great when you only care about having "most recent state".
i.e. for votes we only care about the votes of a proposal "now".
and we wouldn't want to see the state of the votes as of two weeks ago (time = block_height).

that's why i needed to add this new function `new_events_to_process`.
i mentioned to @blushi in our 1-1 today that we should actually convert the `index_class_issuers` process to use this function.
because the same thing applies for class issuers, we only care about the state of the class issuers "now".
the current class issuers indexing process is inefficient for this reason.
we're storing the historical state of the class issuers, and using a flag `latest` to mark the most recent state
i'll open up an issue for this.

another important difference between `new_events_to_process` and `events_to_process` is resiliency (not sure this is the best word).
if you make use of `events_to_process`, you always get all events of a particular type at any point in time.
i.e. if i were to delete a retirement from the retirements table at some old block height, the next time the indexer for retirements runs, it would re-index that retirement without any further manual interactions.
Whereas with `new_events_to_process`, it wouldn't work like that because it only ever pays attention to the events that occurred since the last indexed block height.
in this particular example, you would have to delete all retirements after the initial deleted retirement to get the whole thing re-indexed.

ryanchristo commented on sql/V1_12__votes.sql:
```diff
@@ -0,0 +1,36 @@
+      block_height,
+      tx_idx,
+      msg_idx,
+      TYPE
```
```suggestion
      type
```

ryanchristo commented on index_votes.py:
```diff
@@ -0,0 +1,151 @@
+
+def _index_votes(pg_conn, _client, _chain_num):
+    with pg_conn.cursor() as cur:
+        all_chain_nums = [
+            record[0] for record in gen_records(cur, "select num from chain;")
+        ]
+        max_block_heights = {
+            chain_num: max_block_height
+            for chain_num, max_block_height in gen_records(
+                cur,
+                "select chain_num, MAX(block_height) from votes group by chain_num;",
+            )
+        }
+        logger.debug(f"{all_chain_nums=}")
+        logger.debug(f"{max_block_heights=}")
+        for chain_num in all_chain_nums:
+            if chain_num not in max_block_heights.keys():
+                max_block_heights[chain_num] = 0
+        logger.debug(f"{max_block_heights=}")
+        for chain_num, max_block_height in max_block_heights.items():
+            logger.debug(f"{chain_num=} {max_block_height=}")
```
Also noticing we no longer make use of the `_chain_num` parameter with this addition.

wgwz replied:
> Maybe this is ok for now but this should either be extracted to reuse in other processes or we should rethink how we are running processes, i.e. whether we want to run separate processes for separate chains.

we can figure out a way to extract it. maybe we can do that as a part of refactoring indexing credit class issuers to use the same method.

wgwz replied:
> Not sure if we should address this issue (#33) separately. Any reason for adding it to this pull request?

it’s not really added to this PR. it’s just that this PR will index votes in a forward looking way. if we want to index multiple chains in the same processes, this process will still work for that. (as do all the other processes we’ve written to date).

ryanchristo replied:
> it’s not really added to this PR. it’s just that this PR will index votes in a forward looking way

It's not included in the other processes, e.g. `index_proposals.py`, `index_class_issuers`, etc. We are adding a loop here that does not exist in the other processes, and which seems more relevant to #33 than #22. This pull request is already addressing two issues. It would be nice to focus on one issue at a time.

> if we want to index multiple chains in the same processes, this process will still work for that

As mentioned in the review comment, I'm not sure one process for all chains is the best approach. It might be easier to run multiple instances of the process, one for each chain. We have a single environment variable for a a regen node. It is  diverging from the current architecture and its not required for indexing votes as far as I can tell.

> this is an unused parameter for all the indexing processes so far i.e.

Why was it added in the first place? Is there a future use case for it?

wgwz replied:
> Why was it added in the first place? Is there a future use case for it?

I honestly forgot why I put it there...

> Maybe it could be useful if we want to run multiple instances of the process for each chain?

But yes indeed, it can be useful for this.

> Adding the loop on max_block_heights.items() is needed for the new new_events_to_process logic. But the previous loop on all_chain_nums could potentially be avoided if we use the _chain_num param instead and distribute this logic into different processes.

Yes, I think a good follow up task will be to simplify this as you've described here:

https://github.com/regen-network/indexer/issues/44

This is also something that @ryanchristo and I discussed last week, so good that we are all in agreement it's a good thing to do.

ryanchristo reviewed (commented):
Should we address the multiple chains issue separately? I think there is more work to be done and I'm not sure one process for all chains is the best approach.

ryanchristo commented:
> the way all the indexing processes are set up, is such that if we were to index multiple chains

We have one environment variable for a single regen node, meaning yes it supports multiple chains but only if you run multiple instances, which I do think is the right approach. Adding a loop in the `index_votes.py` script to support multiple chains seems to diverge from the existing architecture.

ryanchristo reviewed (approved):
Per discussion, I think we should consider other solutions to the voter address not being available in the event as well as how we configure and support multiple chains. We should also make sure that we smooth out any inconsistencies now that we are introducing a new approach to how we index votes (i.e. different than how we index class issuers).

I tested a single chain. single proposal, and multiple votes with #41. Not necessarily thorough testing. You can see the logs for each service (i.e. db, indexer, etc.) separately [here](https://github.com/regen-network/indexer/actions/runs/6099955306/job/16553243546).

Nice work overall! Not an easy set of problems to solve.
